api/service/chat.service.go

```diff
@@ -0,0 +1,189 @@
+	chat := domain.NewChat(buyerID, sellerID, sender, message)
+	if err := s.chatRepo.CreateChat(ctx, chat); err != nil {
+		return nil, err
+	}
+
+	s.notifySubscribedSeller(sellerID, chat)
+	return chat, nil
+}
+
+func (s *chatServiceImpl) SendMessageBySeller(ctx context.Context, buyerID int64, sellerID int64, sender domain.UserType, message string) (*domain.Chat, error) {
+	chat := domain.NewChat(buyerID, sellerID, sender, message)
+	if err := s.chatRepo.CreateChat(ctx, chat); err != nil {
+		return nil, err
+	}
+
+	s.notifySubscribedBuyer(buyerID, chat)
+	return chat, nil
+}
+
+func (s *chatServiceImpl) PollMessageByBuyer(ctx context.Context, buyerID int64, sellerID int64, chatID int64) ([]*domain.Chat, error) {
+	latestChatTime := time.Time{}
+	if chatID != -1 {
+		chat, err := s.chatRepo.FindChatByID(ctx, chatID)
+		if err != nil {
+			return nil, err
+		}
+		if chat.BuyerID != buyerID || chat.SellerID != sellerID {
+			return nil, ErrChatNotBelongToOwner
+		}
+		latestChatTime = chat.CreatedAt
+	}
+
+	newChats, err := s.chatRepo.FindLatestChatsByBuyerIDAndSellerID(ctx, buyerID, sellerID, latestChatTime)
+	if err != nil {
+		return nil, err
+	}
+	if len(newChats) > 0 {
+		return newChats, nil
+	}
+
+	newChan := s.addSubscribedBuyers(buyerID)
+	defer s.removeSubscribedBuyer(buyerID, newChan)
+
+	select {
+	case newChat := <-newChan:
+		return newChat, nil
+	case <-time.After(60 * time.Second):
+		return nil, nil
+	}
+}
+
+func (s *chatServiceImpl) PollMessageBySeller(ctx context.Context, buyerID int64, sellerID int64, chatID int64) ([]*domain.Chat, error) {
+	latestChatTime := time.Time{}
+	if chatID != -1 {
+		chat, err := s.chatRepo.FindChatByID(ctx, chatID)
+		if err != nil {
+			return nil, err
+		}
+		if chat.BuyerID != buyerID || chat.SellerID != sellerID {
+			return nil, ErrChatNotBelongToOwner
+		}
+		latestChatTime = chat.CreatedAt
+	}
+
+	newChats, err := s.chatRepo.FindLatestChatsByBuyerIDAndSellerID(ctx, buyerID, sellerID, latestChatTime)
+	if err != nil {
+		return nil, err
+	}
+	if len(newChats) > 0 {
+		return newChats, nil
+	}
+
+	newChan := s.addSubscribedSellers(sellerID)
+	defer s.removeSubscribedSeller(sellerID, newChan)
+
+	select {
+	case newChat := <-newChan:
+		return newChat, nil
+	case <-time.After(60 * time.Second):
+		return nil, nil
+	}
+}
+
+func (s *chatServiceImpl) notifySubscribedBuyer(buyerID int64, chat *domain.Chat) {
+	s.Lock()
+	defer s.Unlock()
+
+	for subscribedBuyer := range s.subscribedBuyers[buyerID] {
+		subscribedBuyer <- []*domain.Chat{chat}
+	}
+}
+
+func (s *chatServiceImpl) notifySubscribedSeller(sellerID int64, chat *domain.Chat) {
+	s.Lock()
+	defer s.Unlock()
+
+	for subscribedSeller := range s.subscribedSellers[sellerID] {
+		subscribedSeller <- []*domain.Chat{chat}
+	}
+}
+
+func (s *chatServiceImpl) addSubscribedBuyers(buyerID int64) chan []*domain.Chat {
```

Comment:
isnt this already enough?

```suggestion
func (s *chatServiceImpl) addSubscribedBuyers(buyerID int64) chan *domain.Chat {
```

Reply:
เด้ว ไม่เหมือนที่คุยไว้
type เราคุยไว้ว่าเป็น
`map[buyerID][]chan *Chat` ป้ะ (มีหลายแนลไว้ส่งให้คลายเครื่อง)
อันนี้มันเป็น 
`map[buyerID]chan []*Chat`